File: Exercicio08.java
import java.io.*;
import java.text.SimpleDateFormat;
import java.util.*;

class Show {
    // atributos
    private String show_id;
    private String type;
    private String title;
    private String director;
    private String[] cast;
    private String country;
    private Date date_added;
    private int release_year;
    private String rating;
    private String duration;
    private String[] listed_in;

    public Show() { // contrutor padrão
        this.show_id = this.type = this.title = this.director = this.country = this.rating = this.duration = "NaN";
        this.cast = new String[0];
        this.listed_in = new String[0];
        this.date_added = null;
        this.release_year = -1;
    }

    // nos sets e nos gets e sets ja posso tratar o ID arrancando o "s" e
    // tranformando o id em int.

    public String getShow_id() {
        return show_id;
    }

    public void setShow_id(String show_id) {
        this.show_id = show_id;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDirector() {
        return director;
    }

    public void setDirector(String director) {
        this.director = director;
    }

    public String[] getCast() {
        return cast;
    }

    public void setCast(String[] cast) {
        this.cast = cast;
    }

    public String getCountry() {
        return country;
    }

    public void setCountry(String country) {
        this.country = country;
    }

    public Date getDate_added() {
        return date_added;
    }

    public void setDate_added(Date date_added) {
        this.date_added = date_added;
    }

    public int getRelease_year() {
        return release_year;
    }

    public void setRelease_year(int release_year) {
        this.release_year = release_year;
    }

    public String getRating() {
        return rating;
    }

    public void setRating(String rating) {
        this.rating = rating;
    }

    public String getDuration() {
        return duration;
    }

    public void setDuration(String duration) {
        this.duration = duration;
    }

    public String[] getListed_in() {
        return listed_in;
    }

    public void setListed_in(String[] listed_in) {
        this.listed_in = listed_in;
    }

    public void readCSV(String linhaCSV) {
        try {
            List<String> camposSeparados = new ArrayList<>();
            StringBuilder bufferCampo = new StringBuilder();
            boolean dentroAspas = false;

            for (int i = 0; i < linhaCSV.length(); i++) {
                char caractere = linhaCSV.charAt(i);

                if (caractere == '"') {
                    dentroAspas = !dentroAspas;
                } else if (caractere == ',' && !dentroAspas) {
                    camposSeparados.add(bufferCampo.toString().trim());
                    bufferCampo.setLength(0);
                } else {
                    bufferCampo.append(caractere);
                }
            }

            camposSeparados.add(bufferCampo.toString().trim());

            String[] campos = camposSeparados.toArray(new String[0]);

            this.show_id = campos[0];
            this.type = campos[1];
            this.title = campos[2].replace("\"", "");
            this.director = campos[3].isEmpty() ? "NaN" : campos[3];
            this.cast = campos[4].isEmpty() ? new String[] { "NaN" } : campos[4].split(",\\s*");
            Arrays.sort(this.cast);
            this.country = campos[5].isEmpty() ? "NaN" : campos[5];

            try {
                SimpleDateFormat formatoData = new SimpleDateFormat("MMMM dd, yyyy", Locale.US);
                this.date_added = campos[6].isEmpty() ? null : formatoData.parse(campos[6]);
            } catch (Exception e) {
                this.date_added = null;
            }

            this.release_year = campos[7].isEmpty() ? -1 : Integer.parseInt(campos[7]);
            this.rating = campos[8].isEmpty() ? "NaN" : campos[8];
            this.duration = campos[9].isEmpty() ? "NaN" : campos[9];
            this.listed_in = campos[10].isEmpty() ? new String[] { "NaN" } : campos[10].split(",\\s*");
            Arrays.sort(this.listed_in);

        } catch (Exception e) {
            System.out.println("Erro ao ler linha: " + e.getMessage());
        }
    }

    public void imprimir() {
        System.out.print("=> " + show_id + " ## " + title + " ## " + type + " ## " + director + " ## [");
        System.out.print(String.join(", ", cast));
        System.out.print("] ## " + country + " ## ");

        if (date_added != null) {
            SimpleDateFormat Saida = new SimpleDateFormat("MMMM d, yyyy", Locale.US);
            System.out.print(Saida.format(date_added));
        } else {
            System.out.print("NaN");
        }

        System.out.print(" ## " + release_year + " ## " + rating + " ## " + duration + " ## [");
        System.out.print(String.join(", ", listed_in));
        System.out.println("] ##");
    }

    public Show clone() {
        Show copia = new Show();
        copia.show_id = this.show_id;
        copia.type = this.type;
        copia.title = this.title;
        copia.director = this.director;
        copia.cast = this.cast.clone();
        copia.country = this.country;
        copia.date_added = this.date_added;
        copia.release_year = this.release_year;
        copia.rating = this.rating;
        copia.duration = this.duration;
        copia.listed_in = this.listed_in.clone();
        return copia;
    }
}

class ListaDupla {
    private CelulaDupla primeiro;
    private CelulaDupla ultimo;

    public Show get(int index) {
        CelulaDupla atual = primeiro.prox;
        for (int i = 0; i < index; i++) {
            atual = atual.prox;
        }
        return atual.elemento;
    }

    public void set(int index, Show value) {
        CelulaDupla atual = primeiro.prox;
        for (int i = 0; i < index; i++) {
            atual = atual.prox;
        }
        atual.elemento = value;
    }

    public void swap(int i, int j) {
        Show temp = get(i);
        set(i, get(j));
        set(j, temp);
    }

    public int getTamanho() {
        return tamanho();
    }

    /**
     * Construtor da classe que cria uma lista dupla sem elementos (somente no
     * cabeca).
     */
    public ListaDupla() {
        primeiro = new CelulaDupla();
        ultimo = primeiro;
    }

    /**
     * Insere um elemento na primeira posicao da lista.
     * 
     * @param x int elemento a ser inserido.
     */
    public void inserirInicio(Show x) {
        CelulaDupla tmp = new CelulaDupla(x);

        tmp.ant = primeiro;
        tmp.prox = primeiro.prox;
        primeiro.prox = tmp;
        if (primeiro == ultimo) {
            ultimo = tmp;
        } else {
            tmp.prox.ant = tmp;
        }
        tmp = null;
    }

    /**
     * Insere um elemento na ultima posicao da lista.
     * 
     * @param x int elemento a ser inserido.
     */
    public void inserirFim(Show x) {
        ultimo.prox = new CelulaDupla(x);
        ultimo.prox.ant = ultimo;
        ultimo = ultimo.prox;
    }

    /**
     * Remove um elemento da primeira posicao da lista.
     * 
     * @return resp int elemento a ser removido.
     * @throws Exception Se a lista nao contiver elementos.
     */
    public Show removerInicio() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover (vazia)!");
        }

        CelulaDupla tmp = primeiro;
        primeiro = primeiro.prox;
        Show resp = primeiro.elemento;
        tmp.prox = primeiro.ant = null;
        tmp = null;
        return resp;
    }

    /**
     * Remove um elemento da ultima posicao da lista.
     * 
     * @return resp int elemento a ser removido.
     * @throws Exception Se a lista nao contiver elementos.
     */
    public Show removerFim() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover (vazia)!");
        }
        Show resp = ultimo.elemento;
        ultimo = ultimo.ant;
        ultimo.prox.ant = null;
        ultimo.prox = null;
        return resp;
    }

    /**
     * Insere um elemento em uma posicao especifica considerando que o
     * primeiro elemento valido esta na posicao 0.
     * 
     * @param x   int elemento a ser inserido.
     * @param pos int posicao da insercao.
     * @throws Exception Se <code>posicao</code> invalida.
     */
    public void inserir(Show x, int pos) throws Exception {

        int tamanho = tamanho();

        if (pos < 0 || pos > tamanho) {
            throw new Exception("Erro ao inserir posicao (" + pos + " / tamanho = " + tamanho + ") invalida!");
        } else if (pos == 0) {
            inserirInicio(x);
        } else if (pos == tamanho) {
            inserirFim(x);
        } else {
            // Caminhar ate a posicao anterior a insercao
            CelulaDupla i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;

            CelulaDupla tmp = new CelulaDupla(x);
            tmp.ant = i;
            tmp.prox = i.prox;
            tmp.ant.prox = tmp.prox.ant = tmp;
            tmp = i = null;
        }
    }

    /**
     * Remove um elemento de uma posicao especifica da lista
     * considerando que o primeiro elemento valido esta na posicao 0.
     * 
     * @param posicao Meio da remocao.
     * @return resp int elemento a ser removido.
     * @throws Exception Se <code>posicao</code> invalida.
     */
    public Show remover(int pos) throws Exception {
        Show resp;
        int tamanho = tamanho();

        if (primeiro == ultimo) {
            throw new Exception("Erro ao remover (vazia)!");

        } else if (pos < 0 || pos >= tamanho) {
            throw new Exception("Erro ao remover (posicao " + pos + " / " + tamanho + " invalida!");
        } else if (pos == 0) {
            resp = removerInicio();
        } else if (pos == tamanho - 1) {
            resp = removerFim();
        } else {
            // Caminhar ate a posicao anterior a insercao
            CelulaDupla i = primeiro.prox;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;

            i.ant.prox = i.prox;
            i.prox.ant = i.ant;
            resp = i.elemento;
            i.prox = i.ant = null;
            i = null;
        }

        return resp;
    }

    /**
     * Mostra os elementos da lista separados por espacos.
     */
    public void mostrar() {
        System.out.print("[ "); // Comeca a mostrar.
        for (CelulaDupla i = primeiro.prox; i != null; i = i.prox) {
            System.out.print(i.elemento + " ");
        }
        System.out.println("] "); // Termina de mostrar.
    }

    /**
     * Mostra os elementos da lista de forma invertida
     * e separados por espacos.
     */
    public void mostrarInverso() {
        System.out.print("[ ");
        for (CelulaDupla i = ultimo; i != primeiro; i = i.ant) {
            System.out.print(i.elemento + " ");
        }
        System.out.println("] "); // Termina de mostrar.
    }

    /**
     * Procura um elemento e retorna se ele existe.
     * 
     * @param x Elemento a pesquisar.
     * @return <code>true</code> se o elemento existir,
     *         <code>false</code> em caso contrario.
     */
    public boolean pesquisar(Show x) {
        boolean resp = false;
        for (CelulaDupla i = primeiro.prox; i != null; i = i.prox) {
            if (i.elemento == x) {
                resp = true;
                i = ultimo;
            }
        }
        return resp;
    }

    /**
     * Calcula e retorna o tamanho, em numero de elementos, da lista.
     * 
     * @return resp int tamanho
     */
    public int tamanho() {
        int tamanho = 0;
        for (CelulaDupla i = primeiro; i != ultimo; i = i.prox, tamanho++)
            ;
        return tamanho;
    }

    class CelulaDupla {
        public Show elemento;
        public CelulaDupla ant;
        public CelulaDupla prox;

        /**
         * Construtor da classe.
         */
        public CelulaDupla() {
            elemento = null;
            ant = prox = null;
        }

        /**
         * Construtor da classe.
         * 
         * @param elemento Show é inserido na celula.
         */
        public CelulaDupla(Show elemento) {
            this.elemento = elemento;
            this.ant = this.prox = null;
        }
    }
}

// classe main
public class Exercicio08 {

    static ListaDupla lista = new ListaDupla();
    static int tamanhoCatalogo = 0;
    static int[] comparacoes = { 0 };
    static int[] movimentacoes = { 0 };

    public static void quickSortLista(int esq, int dir) {
        if (esq < dir) {
            int pivo = particionarLista(esq, dir);
            quickSortLista(esq, pivo - 1);
            quickSortLista(pivo + 1, dir);
        }
    }

    // Método que particiona o array para o Quick Sort
    public static int particionarLista(int esq, int dir) {
        Show pivo = lista.get(dir);
        int i = esq - 1;

        for (int j = esq; j < dir; j++) {
            Show atual = lista.get(j);

            if (atual.getDate_added().compareTo(pivo.getDate_added()) < 0 ||
                    (atual.getDate_added().compareTo(pivo.getDate_added()) == 0 &&
                            atual.getTitle().compareToIgnoreCase(pivo.getTitle()) < 0)) {
                i++;
                lista.swap(i, j);
                movimentacoes[0]++;
            }
            comparacoes[0]++;
        }

        lista.swap(i + 1, dir);
        movimentacoes[0]++;
        return i + 1;
    }

    // faz um swap
    public static void swap(Show[] array, int i, int j) {
        Show temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    static List<Show> catalogo = new ArrayList<>();

    public static void preencherCatalogo() {
        String caminho = "/tmp/disneyplus.csv";
        try {
            BufferedReader br = new BufferedReader(new FileReader(caminho));
            br.readLine();

            String linha;
            while ((linha = br.readLine()) != null) {
                Show s = new Show();
                s.readCSV(linha);
                catalogo.add(s);
            }

            br.close();
        } catch (IOException e) {
            System.out.println("Erro ao ler arquivo: " + e.getMessage());
        }
    }

    public static boolean isFim(String str) {
        return str.equals("FIM");
    }

    // main principal
    public static void main(String[] args) {
        preencherCatalogo();

        Scanner sc = new Scanner(System.in);
        String entrada;

        // Leitura dos IDs e inserção na lista
        while (!(entrada = sc.nextLine()).equals("FIM")) {
            for (Show s : catalogo) {
                if (s.getShow_id().equals(entrada)) {
                    lista.inserirFim(s.clone());
                    break;
                }
            }
        }

        long inicio = System.currentTimeMillis();

        // Ordenação com QuickSort
        quickSortLista(0, lista.getTamanho() - 1);

        long fim = System.currentTimeMillis();

        // Impressão dos elementos ordenados
        for (int i = 0; i < lista.getTamanho(); i++) {
            lista.get(i).imprimir();
        }

        // Gravação do log
        try {
            BufferedWriter log = new BufferedWriter(new FileWriter("matricula_quicksort.txt"));
            log.write("850847\t" + comparacoes[0] + "\t" + movimentacoes[0] + "\t" + (fim - inicio) + "ms");
            log.close();
        } catch (IOException e) {
            System.out.println("Erro ao gravar log: " + e.getMessage());
        }

        sc.close();
    }
}

File: Exercicio08.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <time.h>

#define MAX_LINE 1024
#define MAX_SHOWS 10000
#define MAX_ITEMS 50

// Struct Date
typedef struct Date {
    char mes[20];
    int dia;
    int ano;
} Date;

// Struct show
typedef struct Show {
    char* show_id;
    char* type;
    char* title;
    char* director;
    char** cast;       
    int castSize;      
    char* country;
    Date date_added;   // pega struct Date
    int release_year;
    char* rating;
    char* duration;
    char** listed_in; 
    int listedInSize;
    char* description;
} Show;

// Struct CelulaDupla
typedef struct CelulaDupla {
    Show* elemento;            
    struct CelulaDupla* prox;   
    struct CelulaDupla* ant;    
} CelulaDupla;

// Variáveis globais do catálogo de shows
Show* catalogo[MAX_SHOWS];
int totalShows = 0;

// Variáveis globais para log do quicksort
int comparacoes = 0;
int movimentacoes = 0;

// Função para criar nova célula dupla
CelulaDupla* novaCelulaDupla(Show* elemento) {
    CelulaDupla* nova = (CelulaDupla*) malloc(sizeof(CelulaDupla));
    nova->elemento = elemento;
    nova->ant = nova->prox = NULL;
    return nova;
}

CelulaDupla* primeiro;
CelulaDupla* ultimo;

// Cria nó cabeca
void start() {
    primeiro = novaCelulaDupla(NULL);
    ultimo = primeiro;
}

// Insere no inicio
void inserirInicio(Show* s) {
    CelulaDupla* tmp = novaCelulaDupla(s);

    tmp->ant = primeiro;
    tmp->prox = primeiro->prox;
    primeiro->prox = tmp;

    if (primeiro == ultimo) {                    
        ultimo = tmp;
    } else {
        tmp->prox->ant = tmp;
    }
    tmp = NULL;
}

// Insere no fim
void inserirFim(Show* s) {
    ultimo->prox = novaCelulaDupla(s);
    ultimo->prox->ant = ultimo;
    ultimo = ultimo->prox;
}

// Remove no inicio
Show* removerInicio() {
    if (primeiro == ultimo) {
        fprintf(stderr, "Erro ao remover (vazia)!\n");
        exit(1);
    }

    CelulaDupla* tmp = primeiro;
    primeiro = primeiro->prox;
    Show* resp = primeiro->elemento;
    tmp->prox = primeiro->ant = NULL;
    free(tmp);
    tmp = NULL;
    return resp;
}

// Remove no fim
Show* removerFim() {
    if (primeiro == ultimo) {
        fprintf(stderr, "Erro ao remover (vazia)!\n");
        exit(1);
    } 
    Show* resp = ultimo->elemento;
    ultimo = ultimo->ant;
    ultimo->prox->ant = NULL;
    free(ultimo->prox);
    ultimo->prox = NULL;
    return resp;
}

// Calcula o da lista
int tamanho() {
    int tamanho = 0; 
    CelulaDupla* i;
    for(i = primeiro; i != ultimo; i = i->prox, tamanho++);
    return tamanho;
}

// Insere um elemento 
void inserir(Show* s, int pos) {
    int tam = tamanho();

    if(pos < 0 || pos > tam){
        fprintf(stderr, "Erro ao inserir (posicao %d/%d invalida)!\n", pos, tam);
        exit(1);
    } else if (pos == 0){
        inserirInicio(s);
    } else if (pos == tam){
        inserirFim(s);
    } else {
        // Caminhar ate a posicao anterior a insercao
        CelulaDupla* i = primeiro;
        int j;
        for(j = 0; j < pos; j++, i = i->prox);

        CelulaDupla* tmp = novaCelulaDupla(s);
        tmp->ant = i;
        tmp->prox = i->prox;
        tmp->ant->prox = tmp->prox->ant = tmp;
        tmp = i = NULL;
    }
}

// Remove um elemento 
Show* remover(int pos) {
    Show* resp;
    int tam = tamanho();

    if (primeiro == ultimo){
        fprintf(stderr, "Erro ao remover (vazia)!\n");
        exit(1);
    } else if(pos < 0 || pos >= tam){
        fprintf(stderr, "Erro ao remover (posicao %d/%d invalida)!\n", pos, tam);
        exit(1);
    } else if (pos == 0){
        resp = removerInicio();
    } else if (pos == tam - 1){
        resp = removerFim();
    } else {
        // Caminhar ate a posicao anterior a insercao
        CelulaDupla* i = primeiro->prox;
        int j;
        for(j = 0; j < pos; j++, i = i->prox);

        i->ant->prox = i->prox;
        i->prox->ant = i->ant;
        resp = i->elemento;
        i->prox = i->ant = NULL;
        free(i);
        i = NULL;
    }

    return resp;
}

// Mostra os elementos da lista 
void mostrarListaDupla() {
    CelulaDupla* i;
    int idx = 0;
    for (i = primeiro->prox; i != NULL; i = i->prox) {
       // printf("[%d] ", idx++);
        Show* s = i->elemento;
        printf("=> %s ## %s ## %s ## %s ## [", s->show_id, s->title, s->type, s->director);
        for (int k = 0; k < s->castSize; k++) {
            printf("%s%s", s->cast[k], (k < s->castSize - 1) ? ", " : "");
        }
        printf("] ## %s ## %s %d, %d ## %d ## %s ## %s ## [", s->country, s->date_added.mes, s->date_added.dia,
            s->date_added.ano, s->release_year, s->rating, s->duration);
        for (int k = 0; k < s->listedInSize; k++) {
            printf("%s%s", s->listed_in[k], (k < s->listedInSize - 1) ? ", " : "");
        }
        printf("] ##\n");
    }
}

// Procura show_id no catálogo
Show* buscarNoCatalogo(char* show_id) {
    for (int i = 0; i < totalShows; i++) {
        if (strcasecmp(catalogo[i]->show_id, show_id) == 0) {
            return catalogo[i];
        }
    }
    return NULL;
}

// Funções auxiliares
char* limparString(char* str) {
    if (!str) return strdup("NaN");
    while (isspace(*str)) str++;
    char* end = str + strlen(str) - 1;
    while (end > str && isspace(*end)) end--;
    *(end + 1) = '\0';
    return strdup(str);
}
char* removeAspas(char* str) {
    if (!str) return NULL;
    char* resultado = malloc(strlen(str) + 1);
    int j = 0;
    for (int i = 0; str[i]; i++) {
        if (str[i] != '"') {
            resultado[j++] = str[i];
        }
    }
    resultado[j] = '\0';
    return resultado;
}
char* removeAspasEspacoVazio(char* field) {
    if (!field || strlen(field) == 0) return strdup("NaN");
    char* noQuotes = removeAspas(field);
    char* resultado = limparString(noQuotes);
    free(noQuotes);
    return resultado;
}
char** trataLista(char* str, int* size) {
    char** items = malloc(MAX_ITEMS * sizeof(char*));
    *size = 0;
    if (!str || strlen(str) == 0) {
        items[(*size)++] = strdup("NaN");
        return items;
    }
    char* clear = removeAspas(str);
    char* token = strtok(clear, ",");
    while (token != NULL && *size < MAX_ITEMS) {
        char* item = removeAspasEspacoVazio(token);
        items[(*size)++] = item;
        token = strtok(NULL, ",");
    }
    free(clear);
    //Bubble Sort
    for (int i = 0; i < *size - 1; i++) {
        for (int j = 0; j < *size - i - 1; j++) {
            if (strcasecmp(items[j], items[j + 1]) > 0) {
                char* temp = items[j];
                items[j] = items[j + 1];
                items[j + 1] = temp;
            }
        }
    }
    if (*size == 0) {
        items[(*size)++] = strdup("NaN");
    }
    return items;
}

// Função para separar os campos de uma linha respeitando aspas duplas
int separarCampos(char* linha, char* campos[], int maxCampos) {
    int i = 0;
    char* ptr = linha;
    while (*ptr && i < maxCampos) {
        if (*ptr == '"') {
            ptr++;
            char* start = ptr;
            while (*ptr && !(*ptr == '"' && (*(ptr + 1) == ',' || *(ptr + 1) == '\0'))) {
                ptr++;
            }
            int len = ptr - start;
            campos[i] = malloc(len + 1);
            strncpy(campos[i], start, len);
            campos[i][len] = '\0';
            if (*ptr == '"') ptr++;
            if (*ptr == ',') ptr++;
        } else {
            char* start = ptr;
            while (*ptr && *ptr != ',') ptr++;
            int len = ptr - start;
            campos[i] = malloc(len + 1);
            strncpy(campos[i], start, len);
            campos[i][len] = '\0';
            if (*ptr == ',') ptr++;
        }
        i++;
    }
    return i;
}

// Função para verificar se é FIM
bool isFim(char* str) {
    return strcasecmp(str, "FIM") == 0;
}

void preencherCatalogo() {
    FILE* file = fopen("/tmp/disneyplus.csv", "r");
    if (file == NULL) {
        printf("Erro ao abrir o arquivo!\n");
        return;
    }

    char linha[MAX_LINE];
    fgets(linha, MAX_LINE, file); // Ignora o cabeçalho

    while (fgets(linha, MAX_LINE, file)) {
        linha[strcspn(linha, "\n")] = 0; // Remove o caractere de nova linha

        char* campos[15];
        int fieldCount = separarCampos(linha, campos, 15);

        Show* novoShow = malloc(sizeof(Show));
        novoShow->show_id = removeAspasEspacoVazio(campos[0]);
        novoShow->type = removeAspasEspacoVazio(campos[1]);
        novoShow->title = removeAspasEspacoVazio(campos[2]);
        novoShow->director = removeAspasEspacoVazio(campos[3]);
        novoShow->cast = trataLista(campos[4], &novoShow->castSize);
        novoShow->country = removeAspasEspacoVazio(campos[5]);

        // Trata campo de data
        char* dateStr = removeAspasEspacoVazio(campos[6]);
        if (strcmp(dateStr, "NaN") != 0) {
            char* parte = strtok(dateStr, " ");
            strcpy(novoShow->date_added.mes, parte ? parte : "NaN");
            parte = strtok(NULL, ",");
            novoShow->date_added.dia = parte ? atoi(parte) : -1;
            parte = strtok(NULL, "");
            novoShow->date_added.ano = parte ? atoi(parte) : -1;
        } else {
            strcpy(novoShow->date_added.mes, "March");
            novoShow->date_added.dia = 1;
            novoShow->date_added.ano = 1900;
        }
        free(dateStr);

        novoShow->release_year = (fieldCount > 7 && strlen(campos[7]) > 0) ? atoi(campos[7]) : -1;
        novoShow->rating = removeAspasEspacoVazio(campos[8]);
        novoShow->duration = removeAspasEspacoVazio(campos[9]);
        novoShow->listed_in = trataLista(campos[10], &novoShow->listedInSize);
        novoShow->description = removeAspasEspacoVazio(campos[11]);

        for (int i = 0; i < fieldCount; i++) free(campos[i]);
        catalogo[totalShows++] = novoShow;
    }

    fclose(file);
}

//QuickSort 
// Ordem dos meses para comparar os meses pela posição no ano
const char* mesOrdem[] = {"January", "February", "March", "April", "May", "June", 
                          "July", "August", "September", "October", "November", "December"};

// Função auxiliar para comparar datas e desempatar por title
int compareShows(Show* a, Show* b) {
    int mesA = -1, mesB = -1;
    for (int i = 0; i < 12; i++) {
        if (strcasecmp(a->date_added.mes, mesOrdem[i]) == 0) mesA = i;
        if (strcasecmp(b->date_added.mes, mesOrdem[i]) == 0) mesB = i;
    }
    if (a->date_added.ano != b->date_added.ano)
        return a->date_added.ano - b->date_added.ano;
    if (mesA != mesB)
        return mesA - mesB;
    if (a->date_added.dia != b->date_added.dia)
        return a->date_added.dia - b->date_added.dia;
    return strcasecmp(a->title, b->title);
}

// Algoritmo de ordenação quicksort
void quicksort(Show* array[], int esq, int dir) {
    if (esq < dir) { 
        Show* pivo = array[dir];
        int i = esq - 1;
        for (int j = esq; j < dir; j++) {
            comparacoes++;
            if (compareShows(array[j], pivo) < 0) { 
                i++;
                Show* temp = array[i];
                array[i] = array[j];
                array[j] = temp;
                movimentacoes++;
            }
        }
        Show* temp = array[i + 1];
        array[i + 1] = array[dir];
        array[dir] = temp;
        movimentacoes++;
        int pi = i + 1;
        quicksort(array, esq, pi - 1);
        quicksort(array, pi + 1, dir);
    }
}

int main() {
    preencherCatalogo();

    start(); 

    char entrada[50];
    fgets(entrada, sizeof(entrada), stdin);
    entrada[strcspn(entrada, "\n")] = 0;

    Show* resultados[MAX_SHOWS];
    int tamanhoResultados = 0;

    // ler enquanto não for "FIM"
    while (!isFim(entrada)) {
        Show* show = buscarNoCatalogo(entrada);
        if (show) {
            resultados[tamanhoResultados++] = show;
        }
        fgets(entrada, sizeof(entrada), stdin);
        entrada[strcspn(entrada, "\n")] = 0;
    }

    clock_t inicio = clock();

    // Ordena usando Quicksort 
    quicksort(resultados, 0, tamanhoResultados - 1);

    clock_t fim = clock();

    // Insere os resultados ordenados na lista dupla
    for (int i = 0; i < tamanhoResultados; i++) {
        inserirFim(resultados[i]);
    }

    mostrarListaDupla();

    // Calcula o tempo de execução
    double tempoExecucao = ((double)(fim - inicio)) / CLOCKS_PER_SEC;

    // Gera o arquivo de log 
    FILE* log = fopen("matricula_quicksort.txt", "w");
    if (log) {
        fprintf(log, "850847\t%d\t%d\t%lf\n", comparacoes, movimentacoes, tempoExecucao);
        fclose(log);
    }

    // Libera memória da lista dupla
    CelulaDupla* novaCelula = primeiro;
    while (novaCelula != NULL) {
        CelulaDupla* tmp = novaCelula;
        novaCelula = novaCelula->prox;
        free(tmp);
    }

    // Libera memória dos shows do catálogo
    for (int i = 0; i < totalShows; i++) {
        Show* s = catalogo[i];
        free(s->show_id);
        free(s->type);
        free(s->title);
        free(s->director);
        for (int j = 0; j < s->castSize; j++) free(s->cast[j]);
        free(s->cast);
        free(s->country);
        free(s->rating);
        free(s->duration);
        for (int j = 0; j < s->listedInSize; j++) free(s->listed_in[j]);
        free(s->listed_in);
        free(s->description);
        free(s);
    }

    return 0;
}